great_expectations/agent/message_service/subscriber.py

```diff
@@ -0,0 +1,103 @@
+from functools import partial
+from typing import Callable
+
+import pydantic
+from pika.channel import Channel
+from pika.exceptions import AMQPError, ChannelError
+from pika.spec import Basic, BasicProperties
+
+from great_expectations.agent.message_service.rabbit_mq_client import RabbitMQClient
+from great_expectations.agent.models import Event
+
+OnMessageCallback = Callable[[Event, str], None]
+
+
+class Subscriber:
+    """Manage an open connection to an event stream."""
+
+    def __init__(self, client: RabbitMQClient):
+        """Initialize instance of Subscriber.
+
+        Args:
+            client: RabbitMQClient object with connection and channel attributes.
+        """
+        self.client = client
+
+    def consume(self, queue: str, on_message: OnMessageCallback) -> None:
+        """Subscribe to queue with on_message callback.
+
+        Blocking call which listens to an event stream and invokes on_message
+        with incoming messages.
+
+        Args:
+            queue: name of queue.
+            on_message: callback to be invoked with incoming messages.
+        """
+        # avoid defining _callback_handler inline
+        callback = partial(self._callback_handler, on_message=on_message)
+        try:
+            self.client.channel.basic_consume(queue=queue, on_message_callback=callback)
+            self.client.channel.start_consuming()
+        except (AMQPError, ChannelError) as e:
+            raise SubscriberError from e
+        except KeyboardInterrupt as e:
+            self.client.channel.stop_consuming()
+            raise KeyboardInterrupt from e
+        # user is responsible for calling subscriber.close
+
+    def _callback_handler(
+        self,
+        channel: Channel,
+        method_frame: Basic.Deliver,
+        header_frame: BasicProperties,
+        body: bytes,
+        on_message: OnMessageCallback,
+    ) -> None:
+        """Called by Pika when a message is received.
+
+        Translate message into a known model, obtain any required fields from headers,
+        and pass results into on_message callback.
+
+        Args:
+            channel: the instance of pika.channel.Channel that delivered the message.
+            method_frame: pika object containing context on the delivered message like delivery_tag,
+                consumer_tag, redelivered, exchange, and routing_key.
+            header_frame: pika object containing any header fields, such as correlation_id
+            body: the message body in bytes
+            on_message: the caller-provided callback
+        """
+        try:
+            event = pydantic.parse_raw_as(Event, body)
```

Comment:
```
great_expectations/agent/message_service/subscriber.py:70: error: Need type
annotation for "event"  [var-annotated]
                event = pydantic.parse_raw_as(Event, body)
                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
great_expectations/agent/message_service/subscriber.py:70: error: Argument 1 to
"parse_raw_as" has incompatible type "<typing special form>"; expected
"Type[<nothing>]"  [arg-type]
                event = pydantic.parse_raw_as(Event, body)
```

Reply:
The first error is just saying `event` needs an annotation.
```python
        try:
            event: Event = pydantic.parse_raw_as(Event, body)
```

Reply:
The second error is a bit surprising. Does it actually work at runtime?
Try ignoring the second error for now.